pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -470,262 +559,611 @@ fn storage_sync_loop<
+    let task_name = task.name();
+    match task {
+        SyncTask::Download(new_download_data) => {
+            download_timeline(
+                conf,
+                (storage.as_ref(), &index),
+                current_remote_timeline.as_ref(),
+                sync_id,
+                new_download_data,
+                sync_start,
+                task_name,
+            )
+            .await
+        }
+        SyncTask::Upload(new_upload_data) => {
+            upload_timeline(
+                conf,
+                (storage.as_ref(), &index),
+                current_remote_timeline.as_ref(),
+                sync_id,
+                new_upload_data,
+                sync_start,
+                task_name,
+            )
+            .await;
+            None
+        }
+        SyncTask::DownloadAndUpload(new_download_data, new_upload_data) => {
+            let status_update = download_timeline(
+                conf,
+                (storage.as_ref(), &index),
+                current_remote_timeline.as_ref(),
+                sync_id,
+                new_download_data,
+                sync_start,
+                task_name,
+            )
+            .await;
 
-    let sync_start = Instant::now();
-    let sync_name = task.kind.sync_name();
-    match task.kind {
-        SyncKind::Download(download_data) => {
-            let download_result = download_timeline(
+            upload_timeline(
                 conf,
-                remote_assets.clone(),
-                task.sync_id,
-                download_data,
-                task.retries + 1,
+                (storage.as_ref(), &index),
+                current_remote_timeline.as_ref(),
+                sync_id,
+                new_upload_data,
+                sync_start,
+                task_name,
             )
             .await;
 
-            match download_result {
-                DownloadedTimeline::Abort => {
-                    register_sync_status(sync_start, sync_name, None);
-                    remote_index
+            status_update
+        }
+    }
+}
+
+async fn download_timeline<P, S>(
+    conf: &'static PageServerConf,
+    (storage, index): (&S, &RemoteIndex),
+    current_remote_timeline: Option<&RemoteTimeline>,
+    sync_id: ZTenantTimelineId,
+    new_download_data: SyncData<TimelineDownload>,
+    sync_start: Instant,
+    task_name: &str,
+) -> Option<TimelineSyncStatusUpdate>
+where
+    P: Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+{
+    match download_timeline_layers(storage, current_remote_timeline, sync_id, new_download_data)
+        .await
+    {
+        DownloadedTimeline::Abort => {
+            register_sync_status(sync_start, task_name, None);
+            index
+                .write()
+                .await
+                .set_awaits_download(&sync_id, false)
+                .expect("timeline should be present in remote index");
```

Comment:
I'm just reluctant to panics in general, so not sure if I want to add more.

What if we do otherwise and `error!` here, returning `None`? I've adjusted the code this way.